Add NCCL work sequence number to work info

goldcoderZ opened the pull request:
Summary: Expose sequence number to work info. The number can help applications identify a NCCL work more precisely.

Test Plan: 
1. pytest test/distributed/test_c10d_nccl.py::WorkHookTest::test_on_completion_hook_seq
2. pytest test/distributed/test_c10d_nccl.py::WorkHookTest

Differential Revision: D54180050




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

mrshenli commented on torch/csrc/distributed/c10d/Work.hpp:
```diff
@@ -144,15 +144,18 @@ class TORCH_API Work : public torch::CustomClassHolder {
 struct TORCH_API WorkInfo {
   WorkInfo(
       const OpType& opType,
+      const uint64_t& seq,
```
it's better to pass primitive type by value?

goldcoderZ replied:
> it's better to pass primitive type by value?

yep, thanks!

mrshenli reviewed (commented):
Hey @goldcoderZ, thanks for adding this API. This LGTM. Could you please add a test to this file? Thanks!  https://github.com/pytorch/pytorch/blob/8a32a078563947a6b1de80fb0e8f4234a0b72056/test/distributed/test_c10d_nccl.py

Hey @wconstab @kwen2501  @H-Huang, when you get a chance, could you please help take a look? Thanks!

kwen2501 reviewed (approved):
I am approving because:
- the PR only adds a property binding;
- the `work.getSequencenumber()` call seems to be an existing function.

Next we'd need to consider the relationship between `WorkInfo` (a `TORCH_API` added not long ago) and the trace entries in Flight Recorder, as most of the fields they have are similar. @wconstab

goldcoderZ committed:
Add NCCL work sequence number to work info

Summary: Expose sequence number to work info. The number can help applications identify a NCCL work more precisely.

Test Plan:
```
1. pytest test/distributed/test_c10d_nccl.py::WorkHookTest::test_on_completion_hook_seq
2. pytest test/distributed/test_c10d_nccl.py::WorkHookTest
```

Differential Revision: D54180050

goldcoderZ commented:
@pytorchbot merge

wconstab commented:
what is WorkInfo used by?

goldcoderZ commented:
> what is WorkInfo used by?

It is used by c10d to notify a collateral's completion callback about the comm work's details.

goldcoderZ commented:
@pytorchbot rebase

goldcoderZ commented:
@pytorchbot merge
